### PR TITLE
Use emscripten 1.37.21 on Travis (same as on Circleci)

### LINKED ISSUE #3558: Abort if Emscripten linking results in undefined symbols
From https://github.com/ethereum/solidity/pull/3491#issuecomment-365437236:

```
[100%] Linking CXX executable soljson.js
WARNING:root:not all asm.js optimizations are possible with ALLOW_MEMORY_GROWTH, disabling those. [-Walmost-asm]
warning: unresolved symbol: _ZNSt3__115basic_streambufIcNS_11char_traitsIcEEE5uflowEv
warning: unresolved symbol: _ZNSt3__112basic_stringIcNS_11char_traitsIcEENS_9allocatorIcEEEC1ERKS5_jjRKS4_
warning: unresolved symbol: _ZNSt3__112basic_stringIcNS_11char_traitsIcEENS_9allocatorIcEEEaSERKS5_
...
```

should be a build error.

Use the flag:
```
-s ERROR_ON_UNDEFINED_SYMBOLS=1
```

This is on by default in the CMake Emscripten module which we aren't using.

Ref:
- https://github.com/kripken/emscripten/blob/01fc56a092f8bf83528d6ee8d39db1cd49830f80/src/settings.js#L505-L511
- https://github.com/kripken/emscripten/blob/7f89560101843198787530731f40a65288f6f15f/cmake/Modules/Platform/Emscripten.cmake#L47-L49

### PR DESCRIPTION
**Pull request by axic**
The same version used by the circleci target.

Fixes #3558.

This results in:
- Travis: emscripten changed (1.37.37), boost unchanged (1.57)
- Circleci: emscripten changed (1.37.37), boost unchanged (1.57)

But it fails on cirlceci.

### CODE REVIEW

**Comment by axic** [comment]
Should only merge after the release to avoid any potential problem with using the new emscripten (since no nightlies were created with it yet).

**Comment by axic** [comment]
Strangely enough this works on circleci, but fails on travis:
```
# multiple files can be compiled
missing function: _ZNSt3__112basic_stringIcNS_11char_traitsIcEENS_9allocatorIcEEEC1ERKS5_jjRKS4_
-1
-1
warning: build with  -s DEMANGLE_SUPPORT=1  to link in libcxxabi demangling
```

**Comment by axic** [comment]
Also the build size is different:
Circleci: `Emscripten output size: -rw-r--r-- 1 root root 8219233 Feb 13 22:06 soljson.js`
Travis: `Emscripten output size: -rw-r--r-- 1 root root 8228799 Feb 13 22:58 soljson.js`

(And it took 52 minutes longer for Travis :wink:)

**Comment by axic** [comment]
Should turn those warnings into errors (since not all solc-js test are actually affected by the missing symbols):

```
[100%] Linking CXX executable soljson.js
WARNING:root:not all asm.js optimizations are possible with ALLOW_MEMORY_GROWTH, disabling those. [-Walmost-asm]
warning: unresolved symbol: _ZNSt3__115basic_streambufIcNS_11char_traitsIcEEE5uflowEv
warning: unresolved symbol: _ZNSt3__112basic_stringIcNS_11char_traitsIcEENS_9allocatorIcEEEC1ERKS5_jjRKS4_
warning: unresolved symbol: _ZNSt3__112basic_stringIcNS_11char_traitsIcEENS_9allocatorIcEEEaSERKS5_
...
```

**Comment by aarlt** [comment]
~~@axic what about `-s DEMANGLE_SUPPORT=1`? looks like on circleci it was build with it?~~ It's super strange - the thing is that I spend some hours on the weekend for that. The really strange thing is that it also builds on my machine, with `trzeci/emscripten:sdk-tag-1.37.33-64bit` but that version had exactly the same problem on travis, but additionally to that also a bug in the JavaScript optimizer..

**Comment by axic** [comment]
Just realized that `.travis.yml` contains another docker image reference (to 1.35.4) so perhaps some strange shadowing issue. Not sure why we need a docker pull in travis while using another version in the script.

**Comment by axic** [comment]
Current Travis: `Emscripten output size: -rw-r--r-- 1 root root 7637505 Feb 20 21:47 soljson.js`
Current Circleci: `Emscripten output size: -rw-r--r-- 1 root root 8255543 Feb 20 22:07 soljson.js`
New Travis: `Emscripten output size: -rw-r--r-- 1 root root 8264918 Feb 20 22:13 soljson.js`

But still Travis fails with unresolved symbols while circleci is happy:
```
[100%] Linking CXX executable soljson.js
WARNING:root:not all asm.js optimizations are possible with ALLOW_MEMORY_GROWTH, disabling those. [-Walmost-asm]
warning: unresolved symbol: _ZNSt3__115basic_streambufIcNS_11char_traitsIcEEE5uflowEv
warning: unresolved symbol: _ZNSt3__112basic_stringIcNS_11char_traitsIcEENS_9allocatorIcEEEC1ERKS5_jjRKS4_
warning: unresolved symbol: _ZNSt3__112basic_stringIcNS_11char_traitsIcEENS_9allocatorIcEEEaSERKS5_
warning: unresolved symbol: _ZNSt3__112basic_stringIcNS_11char_traitsIcEENS_9allocatorIcEEE6insertENS_11__wrap_iterIPKcEEc
warning: unresolved symbol: _ZNSt3__112basic_stringIcNS_11char_traitsIcEENS_9allocatorIcEEE9push_backEc
warning: unresolved symbol: _ZNSt3__112basic_stringIcNS_11char_traitsIcEENS_9allocatorIcEEE9__grow_byEjjjjjj
warning: unresolved symbol: _ZNSt3__112basic_stringIcNS_11char_traitsIcEENS_9allocatorIcEEE6appendEPKc
warning: unresolved symbol: _ZNSt3__115basic_streambufIcNS_11char_traitsIcEEE5imbueERKNS_6localeE
warning: unresolved symbol: _ZNSt3__115basic_streambufIcNS_11char_traitsIcEEE9underflowEv
warning: unresolved symbol: _ZNSt3__115basic_streambufIcNS_11char_traitsIcEEE4syncEv
warning: unresolved symbol: _ZNSt3__112basic_stringIcNS_11char_traitsIcEENS_9allocatorIcEEE5eraseEjj
warning: unresolved symbol: _ZNSt3__16localeC1Ev
warning: unresolved symbol: _ZNSt3__112basic_stringIcNS_11char_traitsIcEENS_9allocatorIcEEEC1ERKS5_
warning: unresolved symbol: _ZNSt3__115basic_streambufIcNS_11char_traitsIcEEE6xsgetnEPci
warning: unresolved symbol: _ZNSt3__18ios_base5clearEj
warning: unresolved symbol: _ZNSt3__113basic_istreamIcNS_11char_traitsIcEEED1Ev
warning: unresolved symbol: _ZNSt3__115basic_streambufIcNS_11char_traitsIcEEE9pbackfailEi
warning: unresolved symbol: _ZNSt3__115basic_streambufIcNS_11char_traitsIcEEEC2Ev
warning: unresolved symbol: _ZNSt3__112basic_stringIcNS_11char_traitsIcEENS_9allocatorIcEEE6__initEjc
warning: unresolved symbol: _ZNSt3__115basic_streambufIcNS_11char_traitsIcEEE8overflowEi
warning: unresolved symbol: _ZNKSt3__112basic_stringIcNS_11char_traitsIcEENS_9allocatorIcEEE7compareEPKc
warning: unresolved symbol: _ZNKSt3__121__basic_string_commonILb1EE20__throw_length_errorEv
warning: unresolved symbol: _ZNKSt3__120__vector_base_commonILb1EE20__throw_length_errorEv
warning: unresolved symbol: _ZNKSt3__120__vector_base_commonILb1EE20__throw_out_of_rangeEv
warning: unresolved symbol: _ZNSt3__112basic_stringIcNS_11char_traitsIcEENS_9allocatorIcEEEaSEc
warning: unresolved symbol: _ZNSt3__115basic_streambufIcNS_11char_traitsIcEEE6xsputnEPKci
warning: unresolved symbol: _ZNKSt3__16locale9use_facetERNS0_2idE
warning: unresolved symbol: _ZNSt3__112basic_stringIcNS_11char_traitsIcEENS_9allocatorIcEEED1Ev
warning: unresolved symbol: _ZNSt3__112basic_stringIcNS_11char_traitsIcEENS_9allocatorIcEEE6__initEPKcj
warning: unresolved symbol: _ZNSt3__18ios_base4initEPv
warning: unresolved symbol: _ZNSt3__112basic_stringIcNS_11char_traitsIcEENS_9allocatorIcEEE6appendEjc
warning: unresolved symbol: _ZNSt3__112basic_stringIcNS_11char_traitsIcEENS_9allocatorIcEEE6appendEPKcj
warning: unresolved symbol: _ZNSt3__112basic_stringIcNS_11char_traitsIcEENS_9allocatorIcEEE7reserveEj
warning: unresolved symbol: _ZNSt3__16localeD1Ev
warning: unresolved symbol: _ZNSt13runtime_errorC2ERKNSt3__112basic_stringIcNS0_11char_traitsIcEENS0_9allocatorIcEEEE
warning: unresolved symbol: _ZNSt3__115basic_streambufIcNS_11char_traitsIcEEED2Ev
warning: unresolved symbol: _ZNSt3__112basic_stringIcNS_11char_traitsIcEENS_9allocatorIcEEE6assignEPKc
warning: unresolved symbol: _ZNSt3__16localeC1ERKS0_
warning: unresolved symbol: _ZNSt3__19basic_iosIcNS_11char_traitsIcEEED2Ev
warning: unresolved symbol: _ZNSt3__112basic_stringIcNS_11char_traitsIcEENS_9allocatorIcEEE6assignEjc
warning: unresolved symbol: _ZNSt3__16localeaSERKS0_
warning: unresolved symbol: _ZNKSt3__18ios_base6getlocEv
warning: unresolved symbol: _ZNSt3__112basic_stringIcNS_11char_traitsIcEENS_9allocatorIcEEE6assignEPKcj
warning: unresolved symbol: _ZNSt3__115basic_streambufIcNS_11char_traitsIcEEE9showmanycEv
warning: unresolved symbol: _ZNSt3__113basic_istreamIcNS_11char_traitsIcEEErsERi
warning: unresolved symbol: _ZNSt3__15ctypeIcE2idE
warning: unresolved symbol: _ZNSt3__17collateIcE2idE
warning: unresolved symbol: _ZNSt3__18numpunctIcE2idE
warning: unresolved symbol: _ZTINSt3__115basic_streambufIcNS_11char_traitsIcEEEE
warning: unresolved symbol: _ZTVNSt3__113basic_istreamIcNS_11char_traitsIcEEEE
[100%] Built target soljson
```

**Comment by axic** [comment]
After cleaning the emscripten cache, the new emscripten version fails to build boost on Travis.

**Comment by chriseth** [comment]
Do we still want to include this with 0.4.21?

**Comment by axic** [comment]
No, there's no point forcing this. I'd like to have more testing with the new emscripten with nightlies also it has a >10% size increase and probably will need some tuning once it starts working on travis.

**Comment by chriseth** [comment]
Moved to 0.5.0

**Comment by axic** [comment]
The emscripten we use to ship soljson is 1.35.4 was released on 27 Oct 2015. I wonder what is the risk updating it and what is the risk not updating it.

**Comment by axic** [comment]
Latest emscripten docker has a newer cmake, so the cmake installation step can be removed.

**Comment by axic** [comment]
Travis just abruptly aborts in the middle of the log. The same warnings exists on circle too, but it doesn't abort.

Maybe some kind of out of memory issue?

File: cmake/EthCompilerSettings.cmake [hunk 1]
```diff
@@ -110,6 +110,8 @@ if (("${CMAKE_CXX_COMPILER_ID}" MATCHES "GNU") OR ("${CMAKE_CXX_COMPILER_ID}" MA
 			# Abort if linking results in any undefined symbols
 			# Note: this is on by default in the CMake Emscripten module which we aren't using
 			set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -s ERROR_ON_UNDEFINED_SYMBOLS=1")
+			set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -s VERBOSE=1")
+			set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -s STRICT=1")
```
Comment (axic) [inline cmake/EthCompilerSettings.cmake:114]: @ekpyron Don't fully remember these options, but probably worth keeping it on the other PRs.

**Comment by axic** [comment]
Replaced by #4486.

**Pull request closed without merging**